File: Source/Sql/Parser/Lexer.cpp
#include "Sql/Parser/Lexer.h"

#include <charconv>
#include <string>
#include <system_error>
#include <utility>

#include "Sql/Parser/Token.h"
#include "absl/status/status.h"
#include "absl/strings/ascii.h"
#include "absl/strings/str_cat.h"

namespace spacedb
{
    Lexer::Lexer(std::string_view input) : input_(input)
    {
    }

    bool Lexer::AtEnd() const
    {
        return offset_ >= input_.size();
    }

    char Lexer::Peek() const
    {
        return AtEnd() ? '\0' : input_[offset_];
    }

    char Lexer::Advance()
    {
        return input_[offset_++];
    }

    bool Lexer::IsWhitespace(char value)
    {
        return value == ' ' || value == '\t' || value == '\r' || value == '\n' || value == '\f' || value == '\v';
    }

    bool Lexer::IsIdentifierStart(char value)
    {
        return (value >= 'a' && value <= 'z') || (value >= 'A' && value <= 'Z') || value == '_';
    }

    bool Lexer::IsIdentifierPart(char value)
    {
        return IsIdentifierStart(value) || (value >= '0' && value <= '9');
    }

    void Lexer::SkipWhitespace()
    {
        while (!AtEnd() && IsWhitespace(Peek()))
        {
            Advance();
        }
    }

    bool Lexer::IsDigit(char value)
    {
        return value >= '0' && value <= '9';
    }

    Token Lexer::ScanIdentifier()
    {
        const std::size_t beginOffset = offset_;

        while (!AtEnd() && IsIdentifierPart(Peek()))
        {
            Advance();
        }

        std::string value(input_.substr(beginOffset, offset_ - beginOffset));
        absl::AsciiStrToLower(&value);

        const auto keyword = KeywordFromIdentifier(value);

        if (keyword.has_value())
        {
            return Token{
                .kind = TokenKind::KEYWORD,
                .payload = TokenPayload{keyword.value()},
                .offset = beginOffset,
            };
        }

        return Token{
            .kind = TokenKind::IDENTIFIER,
            .payload = TokenPayload{std::move(value)},
            .offset = beginOffset,
        };
    }

    absl::StatusOr<Token> Lexer::ScanNumber()
    {
        size_t beginOffset = offset_;

        while (!AtEnd() && IsDigit(Peek()))
        {
            Advance();
        }

        bool isFloat = !AtEnd() && Peek() == '.';

        if (isFloat)
        {
            Advance();

            // SQL 不允许 "10." 这类小数点后无数字的字面量
            if (AtEnd() || !IsDigit(Peek()))
            {
                return absl::InvalidArgumentError(absl::StrCat("lexer: digits required after decimal point at ", DescribePosition(beginOffset)));
            }

            while (!AtEnd() && IsDigit(Peek()))
            {
                Advance();
            }
        }

        const std::string_view text = input_.substr(beginOffset, offset_ - beginOffset);

        if (isFloat)
        {
            double value = 0.0;

            const auto [end, error] = std::from_chars(text.data(), text.data() + text.size(), value);

            if (error != std::errc{} || end != text.data() + text.size())
            {
                return absl::InvalidArgumentError(absl::StrCat("lexer: invalid number literal '", text, "' at ", DescribePosition(beginOffset)));
            }

            return Token{
                .kind = TokenKind::NUMBER,
                .payload = TokenPayload{value},
                .offset = beginOffset,
            };
        }

        int64_t value = 0;

        const auto [end, error] = std::from_chars(text.data(), text.data() + text.size(), value);

        if (error == std::errc::result_out_of_range)
        {
            return absl::InvalidArgumentError(absl::StrCat("lexer: integer literal out of range at ", DescribePosition(beginOffset)));
        }

        if (error != std::errc{} || end != text.data() + text.size())
        {
            return absl::InvalidArgumentError(absl::StrCat("lexer: invalid integer literal '", text, "' at ", DescribePosition(beginOffset)));
        }

        return Token{
            .kind = TokenKind::NUMBER,
            .payload = TokenPayload{value},
            .offset = beginOffset,
        };
    }

    absl::StatusOr<Token> Lexer::ScanString()
    {
        // Next() 只有在当前字符为单引号时才调用本函数。
        // 先消费起始单引号
        const std::size_t beginOffset = offset_;

        Advance();

        std::string value;
        while (!AtEnd())
        {
            char current = Advance();
            if (current == '\'')
            {
                return Token{
                    .kind = TokenKind::STRING,
                    .payload = TokenPayload{std::move(value)},
                    .offset = beginOffset,
                };
            }
            value.push_back(current);
        }
        return absl::InvalidArgumentError(absl::StrCat("lexer: unterminated string literal at ", DescribePosition(beginOffset)));
    }

    absl::StatusOr<Token> Lexer::ScanSymbol()
    {
        size_t beginOffset = offset_;

        TokenKind kind;

        switch (Peek())
        {
        case '(':
            kind = TokenKind::OPEN_PAREN;
            break;
        case ')':
            kind = TokenKind::CLOSE_PAREN;
            break;
        case ',':
            kind = TokenKind::COMMA;
            break;
        case ';':
            kind = TokenKind::SEMICOLON;
            break;
        case '*':
            kind = TokenKind::ASTERISK;
            break;
        case '+':
            kind = TokenKind::PLUS;
            break;
        case '-':
            kind = TokenKind::MINUS;
            break;
        case '/':
            kind = TokenKind::SLASH;
            break;
        default:
            return absl::InvalidArgumentError(
                absl::StrCat("lexer: unexpected character '", std::string(1, Peek()), "' at ", DescribePosition(beginOffset)));
        }

        Advance();

        return Token{
            .kind = kind,
            .payload = std::monostate{},
            .offset = beginOffset,
        };
    }

    absl::StatusOr<std::vector<Token>> Lexer::TokenizeAll()
    {
        std::vector<Token> tokens;

        while (true)
        {
            auto token = Next();

            if (!token.ok())
            {
                return token.status();
            }

            bool isEnd = token->kind == TokenKind::END_OF_INPUT;
            tokens.push_back(std::move(*token));

            if (isEnd)
            {
                break;
            }
        }

        return tokens;
    }

    absl::StatusOr<Token> Lexer::Next()
    {
        SkipWhitespace();

        if (AtEnd())
        {
            return Token{
                .kind = TokenKind::END_OF_INPUT,
                .payload = std::monostate{},
                .offset = offset_,
            };
        }

        if (Peek() == '\'')
        {
            return ScanString();
        }

        if (IsDigit(Peek()))
        {
            return ScanNumber();
        }

        if (IsIdentifierStart(Peek()))
        {
            return ScanIdentifier();
        }

        return ScanSymbol();
    }
} // namespace spacedb

File: lexer/lexer.go
package lexer

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

type Lexer struct {
	input  string
	offset int // 当前扫描位置（字节偏移）
}

func New(input string) *Lexer {
	return &Lexer{input: input}
}

// Lex 一次性切出输入的全部 Token（含 EndOfInput），出错立即返回
func Lex(input string) ([]Token, error) {
	return New(input).TokenizeAll()
}

// Next 扫描并返回下一个 Token。
func (l *Lexer) Next() (Token, error) {
	// 跳过空白
	l.skipWhitespace()
	if l.atEnd() {
		return Token{Kind: EndOfInput, Offset: l.offset}, nil
	}

	// 根据当前字符决定：
	// 单引号 → 字符串
	// 数字 → 数字
	// 字母/下划线 → 标识符或关键字
	// 其余 → 符号。
	switch {
	case l.peek() == '\'':
		return l.scanString()
	case isDigit(l.peek()):
		return l.scanNumber()
	case isIdentifierStart(l.peek()):
		return l.scanIdentifier(), nil
	default:
		return l.scanSymbol()
	}
}

// TokenizeAll 循环调用 Next 直到输入结束，返回完整的 Token 列表
func (l *Lexer) TokenizeAll() ([]Token, error) {
	tokens := make([]Token, 0, 8)
	for {
		token, err := l.Next()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, token)
		if token.Kind == EndOfInput {
			return tokens, nil
		}
	}
}

func (l *Lexer) atEnd() bool {
	return l.offset >= len(l.input)
}

func (l *Lexer) peek() byte {
	if l.atEnd() {
		return 0
	}
	return l.input[l.offset]
}

// advance 消费并返回当前字符，把偏移向前推进一位
func (l *Lexer) advance() byte {
	value := l.input[l.offset]
	l.offset++
	return value
}

// skipWhitespace 跳过空白字符
func (l *Lexer) skipWhitespace() {
	remaining := l.input[l.offset:]
	trimmed := strings.TrimLeft(remaining, " \t\r\n\f\v")
	l.offset += len(remaining) - len(trimmed)
}

// scanIdentifier 扫描标识符或关键字
func (l *Lexer) scanIdentifier() Token {
	begin := l.offset
	// 首字符为字母或下划线，后续允许字母、数字、下划线；
	for !l.atEnd() && isIdentifierPart(l.peek()) {
		l.offset++
	}
	text := l.input[begin:l.offset]
	// 匹配关键字表
	if keyword, ok := KeywordFromIdentifier(text); ok {
		return Token{Kind: KeywordKind, Keyword: keyword, Offset: begin}
	}
	// 小写化的标识符
	return Token{Kind: Identifier, Value: strings.ToLower(text), Offset: begin}
}

// scanNumber 扫描数字字面量
func (l *Lexer) scanNumber() (Token, error) {
	begin := l.offset
	for !l.atEnd() && isDigit(l.peek()) {
		l.offset++
	}

	// 小数点后必须还有数字，否则是非法浮点（如 "1."）
	isFloat := !l.atEnd() && l.peek() == '.'
	if isFloat {
		l.offset++
		if l.atEnd() || !isDigit(l.peek()) {
			return Token{}, fmt.Errorf("lexer: digits required after decimal point at %s", DescribePosition(begin))
		}
		for !l.atEnd() && isDigit(l.peek()) {
			l.offset++
		}
	}

	text := l.input[begin:l.offset]
	if isFloat {
		value, err := strconv.ParseFloat(text, 64)
		if err != nil {
			return Token{}, fmt.Errorf("lexer: invalid number literal %q at %s", text, DescribePosition(begin))
		}
		return Token{Kind: Number, Value: value, Offset: begin}, nil
	}

	value, err := strconv.ParseInt(text, 10, 64)
	if err != nil {
		if errors.Is(err, strconv.ErrRange) {
			return Token{}, fmt.Errorf("lexer: integer literal out of range at %s", DescribePosition(begin))
		}
		return Token{}, fmt.Errorf("lexer: invalid integer literal %q at %s", text, DescribePosition(begin))
	}
	return Token{Kind: Number, Value: value, Offset: begin}, nil
}

// scanString 扫描单引号字符串字面量。不支持转义：遇到下一个单引号即结束，
func (l *Lexer) scanString() (Token, error) {
	begin := l.offset
	l.advance() // 跳过开头的单引号

	relativeEnd := strings.IndexByte(l.input[l.offset:], '\'')
	if relativeEnd < 0 {
		l.offset = len(l.input)
		return Token{}, fmt.Errorf("lexer: unterminated string literal at %s", DescribePosition(begin))
	}

	valueStart := l.offset
	l.offset += relativeEnd + 1 // 越过结尾单引号，指向字符串内容之后
	return Token{Kind: String, Value: l.input[valueStart : l.offset-1], Offset: begin}, nil
}

// scanSymbol 扫描单字符符号（括号、逗号、分号、四则运算符）
func (l *Lexer) scanSymbol() (Token, error) {
	begin := l.offset
	var kind Kind
	switch l.peek() {
	case '(':
		kind = OpenParen
	case ')':
		kind = CloseParen
	case ',':
		kind = Comma
	case ';':
		kind = Semicolon
	case '*':
		kind = Asterisk
	case '+':
		kind = Plus
	case '-':
		kind = Minus
	case '/':
		kind = Slash
	case '=':
		kind = Equal
	case '>':
		kind = GreaterThan
	case '<':
		kind = LessThan
	default:
		return Token{}, fmt.Errorf("lexer: unexpected character %q at %s", l.peek(), DescribePosition(begin))
	}
	l.offset++
	return Token{Kind: kind, Offset: begin}, nil
}

func isDigit(value byte) bool {
	return value >= '0' && value <= '9'
}

func isIdentifierStart(value byte) bool {
	return value >= 'a' && value <= 'z' || value >= 'A' && value <= 'Z' || value == '_'
}

func isIdentifierPart(value byte) bool {
	return isIdentifierStart(value) || isDigit(value)
}
